Enhance OIDC authentication flow

sbreker opened the pull request:
Adds two new enhancements for OIDC authentication flow:

1. Add a new setting to control whether local AM authentication is
available when OIDC authentication is in use. If local AM authentication
is disabled, then users will only be able to authenticate via the OIDC
provider. If the new setting is not configured, local AM authentication
is available.

2. Add ability to define more than one OIDC provider in AM. Specific
providers can be chosen using HTTP query params passed to the server
when authenticating.

sbreker committed:
Add OIDC config for testing

Add vars to docker compose file.

sbreker committed:
Enhance OIDC authentication flow

Add a new setting to control whether local AM authentication is
available when OIDC authentication is in use. If local AM authentication
is disabled, then users will only be able to authenticate via the OIDC
provider. If the new setting is not configured, local AM authentication
is available.

Add ability to define more than one OIDC provider in AM. Specific
providers can be chosen using HTTP query params passed to the server
when authenticating.

sbreker committed:
Prevent dashboard polling when not authenticated

sbreker committed:
Add ability to send OIDC logout

Override the mozilla_django_oidc OIDCLogoutView class to allow
Archivematica to support sending an OIDC logout message to OIDC providers
that support it.

sbreker committed:
Allow configuration of endpoints by provider

Allow independent configuration of OIDC endpoints for each configured
provider.

sbreker committed:
Add override CustomOIDCAuthenticationCallbackView

Ensure get_settings() is overridden for the mozilla_django_oidc
OIDCAuthenticationCallbackView class.

replaceafill commented on src/dashboard/src/templates/layout.html:
```diff
@@ -19,7 +19,9 @@
     <script src="{% static 'vendor/underscore.js' %}" type="text/javascript"></script>
     <script src="{% static 'vendor/backbone.js' %}" type="text/javascript"></script>
-    <script src="{% static 'js/status.js' %}" type="text/javascript"></script>
+    {% if user.is_authenticated %}
+      <script src="{% static 'js/status.js' %}" type="text/javascript"></script>
+    {% endif %}
```
Thank you for addressing this :tada:!

For the record, Steve found that the `login` view kept redirecting back to itself after a few seconds which was caused by an attempt to poll data from the MCPServer which is only available to authenticated users.

sbreker replied:
I will not squash this commit - I will keep it as a separate individual commit since it is unrelated to the OIDC changes.

replaceafill commented on src/dashboard/src/components/accounts/backends.py:
```diff
@@ -42,6 +43,49 @@ class CustomOIDCBackend(OIDCAuthenticationBackend):
+            "OIDC_OP_LOGOUT_ENDPOINT",
+        ]:
+            # Retrieve the request object stored in the instance.
+            request = getattr(self, "request", None)
+
+            if request:
+                provider_name = request.session.get("providername")
```
Is it necessary to query the session through all these iterations?

sbreker replied:
Since this session value is the source of record for the current provider, my strategy was to rely on this variable completely rather than make local copies in the oidc derived classes. I thought this was simpler and more clear and avoids the problem of checking if the provider has changed, and comparing it to a stored provider name. This code will only be executed if one of these seven env vars are queried - any others will use the base class get_settings().

Is there a performance problem I am not accounting for with checking a session var?

replaceafill replied:
We plan to address this refactoring after adding some basic tests for the new components.

replaceafill replied:
I misread the code here. These were not iterations, but `if` conditions :facepalm:. Sorry for the noise.

replaceafill commented on src/dashboard/src/components/accounts/backends.py:
```diff
@@ -42,6 +43,49 @@ class CustomOIDCBackend(OIDCAuthenticationBackend):
+
+    def authenticate(self, request, **kwargs):
+        self.request = request
+        self.OIDC_RP_CLIENT_ID = self.get_settings("OIDC_RP_CLIENT_ID")
+        self.OIDC_RP_CLIENT_SECRET = self.get_settings("OIDC_RP_CLIENT_SECRET")
+        self.OIDC_OP_TOKEN_ENDPOINT = self.get_settings("OIDC_OP_TOKEN_ENDPOINT")
+        self.OIDC_OP_USER_ENDPOINT = self.get_settings("OIDC_OP_USER_ENDPOINT")
+        self.OIDC_OP_JWKS_ENDPOINT = self.get_settings("OIDC_OP_JWKS_ENDPOINT")
```
If I understand the `get_settings` method correctly this block is going to try to iterate the 7 attributes these many times?

replaceafill replied:
Sorry Steve. I misread the `if` conditions at the beginning of the `get_settings` methods as `for` loops :facepalm:

replaceafill commented on src/dashboard/src/components/accounts/views.py:
```diff
@@ -193,3 +201,178 @@ def delete(request, id):
+    OIDC client authentication callback HTTP endpoint
+    """
+
+    def get_settings(self, attr, *args):
```
Is the logic in these `get_settings` methods similar enough to be extracted to a common base class?

replaceafill replied:
We plan to address this refactoring after adding some basic tests for the new components.

replaceafill replied:
The callback class has been removed as it seemed unnecessary for the feature.

replaceafill commented on src/dashboard/src/components/accounts/views.py:
```diff
@@ -193,3 +201,178 @@ def delete(request, id):
+
+def get_oidc_logout_url(request):
+    """
+    Constructs the OIDC logout URL for Keycloak.
```
Is this logic Keycloak specific? If not, maybe it's not worth mentioning it in the commment?

replaceafill reviewed (changes requested):
Great progress @sbreker! I've left a few comments and please bear with me that this is early feedback since I'm still getting familiar with the approach :sweat_smile: 

I think my biggest concern at this point is about testing the new code. Automated unit tests would be preferably but even having an option to manually test the new feature would be nice.

I tried testing the feature manually by setting the PR branch and noticed that the settings added to the Compose services are turning the feature on  by default in the development environment which now shows a `Log in with OpenID Connect` link in the authentication view. Clicking on it however points me to a non existing `keycloak` service.

I wonder if we should consider making the feature opt-in. Maybe by adding a separate `docker-compose.keycloak.yml` file that [extends](https://docs.docker.com/compose/multiple-compose-files/extends/) the existing services and adds a fully integrated service that runs a `keycloak` container, similarly to the `oidc-demo` project you once showed me.

Should we also try to document this feature more extensively? The PR description mentions new settings and abilities but doesn't list their names or default values. Would it be more convenient to either use:

* The [OIDC variables](https://github.com/artefactual/archivematica/tree/d6b3869ca742c6be955e4bd34f640d59949941b5/src/dashboard/install#oidc-variables) section of the Dashboard README
* The [OIDC setup](https://www.archivematica.org/en/docs/archivematica-1.16/admin-manual/security/security/#oidc-setup) section of the docs
* Maybe both?

I think some design decisions like uppercasing the secondary provider name should be reflected too.

During my brief testing I also noticed some weird interaction while switching between the Dashboard and the Storage Service UIs. I haven't been able to reproduce it consistently but sometimes after logging out from one service the other service starts giving me CSRF errors. I'll comment again if I'm able to find the right sequence of steps to trigger it.

Finally I'm wondering how the new setting to allow "local authentication" relates to Django's own [authentication heuristics](https://docs.djangoproject.com/en/4.2/topics/auth/customizing/#specifying-authentication-backends) where a user should be able to set up multiple authentication backends in a list. I'll try to do some more research on this and comment back if I see any potential problem.

sbreker commented:
Hi @replaceafill! Thanks for this feedback - I will start addressing these items.

One thing I forgot to mention is the commit that updates the docker compose file will be dropped - the defaults for AM's authentication are not meant to be updated.

I have started crafting some configuration instructions [in this gist](https://gist.github.com/sbreker/2022a1af14adeb665bbe699b62bdce8d).

sbreker commented:
I think I may need you to elaborate on what you are looking for wrt to documentation here: "I think some design decisions like uppercasing the secondary provider name should be reflected too."

The secondary provider name config is case insensitive. Do you think the code should be changed to force it to be specified as uppercase?

sbreker committed:
Update comment in get_oidc_logout_url()

sbreker committed:
Fix settings issue in CustomOIDCLogoutView

Fixes issue where the 'providername' session variable is was not deleted
on logout.

Fixes issue where secondary OIDC endpoint logout path was not being used
when a secondary provider was used.

sbreker committed:
Remove uppercase logic from oidc_auth

Remove automatic conversion of provider names to uppercase from
get_oidc_secondary_providers().

Update the OIDC middleware to convert the secondary provider name to
uppercase before validation.

sbreker committed:
Replace allow_local_auth logic

Removed custom OIDC login view in favour of updating the LOGIN_URL based
on system configuration.

replaceafill committed:
Add integration test for secondary request parameter

sbreker committed:
Update logic adding OIDC Middleware

Change where middleware is added in the list of middleware.

replaceafill committed:
Move middleware after the AuthenticationMiddleware

replaceafill committed:
Adjust OIDC_ALLOW_LOCAL_AUTHENTICATION behavior

replaceafill committed:
Remove custom OIDC callback class

replaceafill committed:
Remove testing data from Compose configuration

replaceafill committed:
Add test for OIDC capture middleware

replaceafill committed:
Add tests for get_oidc_logout_url

sbreker reviewed (commented):
Looks great @replaceafill! :+1:
